msfextract

```diff
@@ -0,0 +1,240 @@
+#!/usr/bin/env ruby
+# -*- coding: binary -*-
+
+class MsfExtractError < StandardError; end
+class HelpError < StandardError; end
+class UsageError < MsfExtractError; end
+
+require 'tmpdir'
+require 'optparse'
+require 'timeout'
+require 'open3'
+require 'nokogiri'
+require 'fileutils'
+
+def run_cmd(cmd)
+  begin
+    stdin, stdout, stderr = Open3.popen3(cmd)
+    return stdout.read + stderr.read
+  rescue Errno::ENOENT
+    return nil
+  end
+end
+
+def parse_manifest(manifest_file)
+  File.open(manifest_file, "rb"){|file|
+    data = File.read(file)
+    return Nokogiri::XML(data)
+  }
+end
+
+def extract_data(java_file, text_file)
+  file_data = File.read("#{java_file}")
+  mid = file_data.split("[]{")
+  bytes = mid[1].split("}")
+  bytes[0]
+
+  File.open("#{text_file}", 'wb') do |f|
+    f.write(bytes[0])
+  end
+
+  $stderr.puts "[+] Saved as: #{text_file}"
+end
+
+def check_tools
+  # CHECK IF ALL TOOLS ARE INCLUDED BEFORE STARTING
+  unzip_test = run_cmd("unzip")
+  unless unzip_test != nil
+    raise RuntimeError, "unzip command not found."
+  end
+
+  d2j_test = run_cmd("d2j-dex2jar")
+  unless d2j_test != nil
+    raise RuntimeError, "d2j-dex2jar command not found."
+  end
+
+  java_test = run_cmd("java")
+  unless java_test != nil
+    raise RuntimeError, "java command not found."
+  end
+
+  apktool_test = run_cmd("apktool")
+  unless apktool_test != nil
+    raise RuntimeError, "apktool command not found."
+  end
```

Comment:
You could really put all this in a loop like
```suggestion
["unzip", "d2j-dex2jar"...].each do |tool|
  raise RuntimeError, "#{tool} command not found." if run_cmd(tool).nil?
end
```

Reply:
There's also a cross platform [`which`](https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/util/helper.rb#L9) implementation in `lib/msf/util/helper.rb` that this should use instead of trying to run the command.

Reply:
Might be nice to also convert these checks to do lighter checks either using `which/where` to eval for packages or using the `Rex` libraries' detection code such as `Rex::FileUtils.find_full_path()`.  Here is an example: https://github.com/rapid7/metasploit-framework/blob/81b978964ab64cb250418b7f0831b1ba20d3e4d7/lib/metasploit/framework/password_crackers/cracker.rb#L276-L284